### PR TITLE
Refactor api/client/build.go for trusted pulls

### PR DESCRIPTION
**Pull request by jlhawn**
Better handling of errors on context uploads:
- performs notary lookups *before* context upload begins

Separated handling of context and Dockerfile processing.

### CODE REVIEW

**Comment by dmcgowan** [comment]
I'll get this added in but won't have time to review it until a bit later.

One quick note on the util package name, `builder/builderutil` would be a more idiomatic package name than `builder/util`

**Comment by jlhawn** [comment]
isn't `builder/builderutil` even less idiomatic because it stutters? When importing it from other packages though, I do use it as `builderutil` because the name `util` is often overloaded:

```go
import (
	builderutil "github.com/docker/docker/builder/util"
)
```

File: api/client/build.go [hunk 1]
```diff
@@ -72,211 +72,101 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
...
+	}
 
-		}()
+	// If .dockerignore mentions .dockerignore or the Dockerfile
+	// then make sure we send both files over to the daemon
+	// because Dockerfile is, obviously, needed no matter what, and
+	// .dockerignore is needed to know if either one needs to be
+	// removed.  The deamon will remove them for us, if needed, after it
+	// parses the Dockerfile.
+	keepThem1, _ := fileutils.Matches(".dockerignore", excludes)
```
Comment (stevvooe) [inline api/client/build.go]: Check errors here.

File: api/client/build.go [hunk 2]
```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	return err
 }
 
-var (
-	dockerfileFromLinePattern        = regexp.MustCompile(`(?i)^FROM[ \f\r\t\v]+(?P<image>[^ \f\r\t\v\n#]+)`)
-	dockerfileLineContinutionPattern = regexp.MustCompile(`\\[ \t]*$`)
-)
+func getDockerfileRelPath(dockerfileName *string, baseDir string) (err error) {
```
Comment (stevvooe) [inline api/client/build.go]: Document this function and explain why `dockerfileName` is a pointer.

File: api/client/build.go [hunk 3]
```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
...
+	}
+
+	if *dockerfileName, err = filepath.Rel(absBase, dockerfileAbspath); err != nil {
+		return fmt.Errorf("unable to get relative Dockerfile path: %v", err)
+	}
+
+	return nil
+}
+
+func createDockerfileFromReader(r io.Reader, filename string) error {
```
Comment (stevvooe) [inline api/client/build.go]: This should be called `writeFile or something like that. Alternatively, this can be completely replaced with `ioutil.WriteFile`.

File: api/client/build.go [hunk 4]
```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
...
+	}
+
+	return nil
+}
+
+// As a special case, 'docker build -' will build from either an empty context
+// with the contents of stdin as a Dockerfile, or a tar archive of context from
+// stdin. getContextFromStdin will create a local temporary directory which
+// contains this context and/or Dockerfile.
+func (cli *DockerCli) getContextFromStdin(dockerfileName *string) (tempDir string, err error) {
```
Comment (stevvooe) [inline api/client/build.go]: Make sure to follow Go documentation guidelines for these functions:

```
// getContextFromStdin does blah blah blah...
```

File: api/client/build.go [hunk 5]
```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
...
+	}
+
+	if err := archive.Untar(buf, tempDir, nil); err != nil {
+		return "", fmt.Errorf("unable to extract stdin to temporary context direcotry: %v", err)
+	}
+
+	return tempDir, getDockerfileRelPath(dockerfileName, tempDir)
+}
+
+func (cli *DockerCli) getContextFromGitURL(gitURL string, dockerfileName *string) (tempDir string, err error) {
```
Comment (stevvooe) [inline api/client/build.go]: I'm a little worried that this struct method set is sprawling. There is some sort of context here that is being created as a directory. This seems orthogonal to the concerns of the `DockerCli` type. These operations are related to building.

File: api/client/build.go [hunk 6]
```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
...
+}
+
+var dockerfileFromLinePattern = regexp.MustCompile(`(?i)^FROM[ \f\r\t\v]+(?P<image>[^ \f\r\t\v\n#]+)`)
 
 // makeTrustedDockerfileFrom rewrites a Dockerfile read from the given input
 // by resolving images in "FROM <image>" instructions to a trusted digest
 // reference.
-func (cli *DockerCli) makeTrustedDockerfileFrom(input io.Reader, approximateSize int64) (*bytes.Buffer, error) {
-	scanner := bufio.NewScanner(input)
+func (cli *DockerCli) makeTrustedDockerfileFrom(dockerfileName string) (newDockerfile *bytes.Buffer, err error) {
```
Comment (stevvooe) [inline api/client/build.go]:
```
func pinReferences(dst io.Writer, src io.Reader, pin func(repo, ref string) (string, error)) error
```

File: api/client/build.go [hunk 7]
```diff
@@ -427,20 +488,64 @@ func (cli *DockerCli) makeTrustedDockerfileFrom(input io.Reader, approximateSize
...
+		if _, err := fmt.Fprintln(buf, line); err != nil {
 			return nil, err
 		}
 	}
 
-	return dockerfileBuf, scanner.Err()
+	return buf, scanner.Err()
+}
+
+func trustedDockerfileTarWrapper(inputTarStream io.Reader, newDockerfile *bytes.Buffer, dockerfileName string) io.Reader {
```
Comment (stevvooe) [inline api/client/build.go]: Might need to return an `io.ReadCloser` here.

File: api/client/build.go [hunk 8]
```diff
@@ -427,20 +488,64 @@ func (cli *DockerCli) makeTrustedDockerfileFrom(input io.Reader, approximateSize
...
+				return
+			}
+			if err != nil {
+				pipeWriter.CloseWithError(err)
+				return
+			}
+
+			var content io.Reader = tarReader
+
+			if hdr.Name == dockerfileName {
```
Comment (stevvooe) [inline api/client/build.go]: Is the name a strong enough check? I think it is but this might be worth a note here that indicates it is and why.

File: builder/util/util.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+const acceptableRemoteMIME = `(?:application/(?:(?:x\-)?tar|octet\-stream|((?:x\-)?(?:gzip|bzip2?|xz)))|(?:text/plain))`
+
+var mimeRe = regexp.MustCompile(acceptableRemoteMIME)
+
+func selectAcceptableMIME(ct string) string {
+	return mimeRe.FindString(ct)
+}
+
+func HandleJsonArgs(args []string, attributes map[string]bool) []string {
```
Comment (stevvooe) [inline builder/util/util.go]: Please document this.

File: builder/util/util.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
+package util
```
Comment (stevvooe) [inline builder/util/util.go]: This does not warrant an extra package. Just leave this is an extra file in the builder package.

File: builder/util/util.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+// InspectResponse looks into the http response data at r to determine whether its
+// content-type is on the list of acceptable content types for remote build contexts.
+// This function returns:
+//    - a string representation of the detected content-type
+//    - an io.Reader for the response body
+//    - an error value which will be non-nil either when something goes wrong while
+//      reading bytes from r or when the detected content-type is not acceptable.
+func InspectResponse(ct string, r io.ReadCloser, clen int) (string, io.ReadCloser, error) {
```
Comment (stevvooe) [inline builder/util/util.go]: How is this different than [`http.DetectContentType`](http://golang.org/pkg/net/http/#DetectContentType)?

I do see why, but perhaps it should follow that function interface.

**Comment by stevvooe** [comment]
For the most part, this looks good. There are few structural issues, but they may be outside the scope of this PR.

File: builder/util/util.go (hunk 10 above)
Comment (jlhawn) [inline builder/util/util.go]: I didn't want to go through the headache of getting the `"github.com/docker/docker/builder"` package to compile for the client binary distribution on all platforms. It's easier to put it in a separate package.

File: api/client/build.go (hunk 1 above)
Comment (jlhawn) [inline api/client/build.go]: These errors are checked in another call to `ValidateContextDirectory(contextDir, excludes)` below. I'll move that to come before this and mention it in the comment. I also didn't write this code, it's just been unindented.

File: api/client/build.go (hunk 2 above)
Comment (jlhawn) [inline api/client/build.go]: I'm redoing most of these so the pointer type wont be an issue. I'm also going through and adding comments to each of the added functions.

File: api/client/build.go (hunk 3 above)
Comment (jlhawn) [inline api/client/build.go]: `ioutil.WriteFile` takes `[]byte` rather than `io.Reader` otherwise I'd use that.

File: api/client/build.go (hunk 4 above)
Comment (jlhawn) [inline api/client/build.go]: I just took this comment block from the previous `if/else` branch that this code was in. I'm going through now and adding/updating all of the comments.

File: builder/util/util.go (hunk 10 above)
Comment (stevvooe) [inline builder/util/util.go]: Ick. Can you put a note that this should go back into that package if that problem is solved?

File: api/client/build.go (hunk 5 above)
Comment (jlhawn) [inline api/client/build.go]: yeah, some of these don't even need access to the `cli` object. I should be able to factor out the ones that do and make these all just functions within the file.

File: api/client/build.go (hunk 3 above)
Comment (stevvooe) [inline api/client/build.go]:
```
func writeFile(filename string, rd io.Reader) error {
    p, err := ioutil.ReadAll(rd)
    if err != nil { return err }

    return ioutil.WriteFile(filename, p)
}
```

It took way to long to figure out what this function was actually doing.

**Comment by jlhawn** [comment]
I've removed the changes to the `builder` package, including the `builder/util` package. The current commit is just refactoring. I'll add the Dockerfile rewriting in a second commit.

**Comment by stevvooe** [comment]
LGTM

**Comment by jlhawn** [comment]
second commit is up.

**Comment by stevvooe** [comment]
LGTM

I don't see any obvious problems, except the overuse of propositions in function names. ;)

**Comment by jlhawn** [comment]
@stevvooe how about:

```go
func fromFromToFromToBeforeClose() error
```

**Comment by stevvooe** [comment]
:smiley_cat:
